src/main/java/org/prebid/server/bidder/metax/MetaxBidder.java

```diff
@@ -0,0 +1,158 @@
+    private static final String AD_UNIT_MACRO = "{{adUnit}}";
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public MetaxBidder(String endpointUrl,
+                       JacksonMapper mapper) {
+
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpMetax extImpMetax = parseImpExt(imp);
+                httpRequests.add(BidderUtil.defaultRequest(prepareBidRequest(request, imp),
+                        resolveEndpoint(extImpMetax),
+                        mapper));
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        return Result.of(httpRequests, errors);
+    }
+
+    private ExtImpMetax parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), METAX_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private static BidRequest prepareBidRequest(BidRequest bidRequest, Imp imp) {
+        return bidRequest.toBuilder()
+                .imp(Collections.singletonList(modifyImp(imp)))
+                .build();
+    }
+
+    private static Imp modifyImp(Imp imp) {
+        final Banner banner = imp.getBanner();
+        final Integer width = banner != null ? banner.getW() : null;
+        final Integer height = banner != null ? banner.getH() : null;
+        if (width != null && height != null) {
+            return imp;
+        }
+
+        final List<Format> formats = banner != null ? banner.getFormat() : null;
+        if (CollectionUtils.isEmpty(formats)) {
+            return imp;
+        }
+
+        final Format firstFormat = formats.getFirst();
+        return imp.toBuilder()
+                .banner(banner.toBuilder()
+                        .w(zeroIfFormatMeasureNull(firstFormat, Format::getW))
+                        .h(zeroIfFormatMeasureNull(firstFormat, Format::getH))
+                        .build())
+                .build();
+    }
+
+    private static Integer zeroIfFormatMeasureNull(Format format, Function<Format, Integer> measureExtractor) {
+        return ObjectUtils.defaultIfNull(ObjectUtil.getIfNotNull(format, measureExtractor), 0);
+    }
+
+    private String resolveEndpoint(ExtImpMetax extImpMetax) {
+        return endpointUrl
+                .replace(PUBLISHER_ID_MACRO, String.valueOf(extImpMetax.getPublisherId()))
+                .replace(AD_UNIT_MACRO, String.valueOf(extImpMetax.getAdUnit()));
```

Comment:
I know that both fields are mandatory and the following case should not be possible, but `String.valueOf()` returns `"null"` if the value is null. I'm not sure it's a good output in any case, perhaps makes sense to use `StringUtils.defaultString(...)` instead

Reply:
StringUtils.defaultString(...) will not work with integers, so updated to approach with optionals